# Car Listing Analysis #

This project is designed to analyse used car listings from eBay Kleinanzeigen, a classifieds section of the German eBay website.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [4]:
# Checking to see if the import worked
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
# This gives us an idea of what columns are missing data. We can see vehicleType, gearbox, model, fuelType, and notRepairedDamage have significant null objects.

autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataframe has several columns with large swathes of missing data - damage, fueltype, etc. Some of these may overlap in terms of information given, or it may be necessary to clear off every row with a null depending on what we are looking for.

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
# Let's clean these columns up with some different names.

auto_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type',
                'registration_year','gearbox','powerPS','model','odometer','registration_month',
                'fuel_type','brand','unrepaired_damage','ad_created','nr_of_pictures', 'postal_code',
                'last_seen']

In [9]:
autos.columns = auto_columns

In [10]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [12]:
autos.head() # Checking to see if the columns look ok.

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuel_type  \
0    manuell      158  andere  150,000km                   3       lpg   
1  automatik      286     7er  150,000km                   6    benzin   
2    manuell      102    golf   70,000km                   7    benzin   
3  automatik       71  fortwo   70,000km                   6    benzin   
4    manuell        0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [13]:
# Now we're ready to look at some data. Describe isn't picking up on price. Why?

autos.describe()

registration_year       powerPS  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

In [20]:
# Looks like price is a string with commas and dollar signs.

autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [15]:
# Price is currently listed in dollars and has commas. It's also a string. We need to clean it up and convert to an int to be able to get more information out of it.

autos["price"] = autos["price"].str.replace('$','')

autos["price"] = autos["price"].str.replace(',','')


autos["price"] = autos["price"].astype(int)

In [17]:
# This looks better.

autos["price"]

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int64

In [18]:
# That's better. 

autos.describe()

price  registration_year       powerPS  registration_month  \
count  5.000000e+04       50000.000000  50000.000000        50000.000000   
mean   9.840044e+03        2005.073280    116.355920            5.723360   
std    4.811044e+05         105.712813    209.216627            3.711984   
min    0.000000e+00        1000.000000      0.000000            0.000000   
25%    1.100000e+03        1999.000000     70.000000            3.000000   
50%    2.950000e+03        2003.000000    105.000000            6.000000   
75%    7.200000e+03        2008.000000    150.000000            9.000000   
max    1.000000e+08        9999.000000  17700.000000           12.000000   

       nr_of_pictures   postal_code  
count         50000.0  50000.000000  
mean              0.0  50813.627300  
std               0.0  25779.747957  
min               0.0   1067.000000  
25%               0.0  30451.000000  
50%               0.0  49577.000000  
75%               0.0  71540.000000  
max               0.0  99998.000000

In [21]:
# Doing the same for odometer data.

autos["odometer"] = autos["odometer"].str.replace('km','')

autos["odometer"] = autos["odometer"].str.replace(',','')

autos["odometer"] = autos["odometer"].astype(int)

In [22]:
autos.describe()

price  registration_year       powerPS       odometer  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [24]:
# Add some units to the odomoter column name.

autos.rename(columns={'odometer' : 'odometer_km'}, inplace=True)


In [25]:
autos.describe()

price  registration_year       powerPS    odometer_km  \
count  5.000000e+04       50000.000000  50000.000000   50000.000000   
mean   9.840044e+03        2005.073280    116.355920  125732.700000   
std    4.811044e+05         105.712813    209.216627   40042.211706   
min    0.000000e+00        1000.000000      0.000000    5000.000000   
25%    1.100000e+03        1999.000000     70.000000  125000.000000   
50%    2.950000e+03        2003.000000    105.000000  150000.000000   
75%    7.200000e+03        2008.000000    150.000000  150000.000000   
max    1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

In [35]:
# Time to start exploring odometer data.

odo = autos["odometer_km"]
uniq = odo.unique().shape
print(uniq)
odo_counts = odo.value_counts()
odo_counts.sort_index(ascending=True)

(13,)


5000        760
10000       245
20000       757
30000       777
40000       814
50000      1009
60000      1153
70000      1214
80000      1412
90000      1733
100000     2101
125000     5037
150000    31212
Name: odometer_km, dtype: int64

In [36]:
odo.describe()

count     48224.000000
mean     125919.148142
std       39543.339640
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Our odometer data looks fairly reasonable, though clearly 150,000 plus is 150,000+ and 5,000 is <5,000. Checking through this immediately doesn't yield anything non-obvious, though it seems the vast majority of used cars have higher odometer readings - this seems self-evident, though. Unfortunately there's no granularity of data above 150,000 kms,

In [28]:
# Let's do the same for pricing data.

price = autos["price"]
price_uniq = price.unique().shape
print(price_uniq)
price_counts = price.value_counts()
price_counts.sort_index(ascending=False).head(15)

(2357,)


99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [30]:
price_counts.sort_index(ascending=True).head(25)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
40       6
45       4
47       1
49       4
50      49
Name: price, dtype: int64

There are a significant number of price points (to be expected), but some are clearly incorrect. Everything above 350,000 looks like data put in to simply fill in a required field. Likewise, we don't really expect 1421 free cars - this is likely data put in to fit a requirement. Let's cap the price data at 350,000, and make minimum car cost $100.

In [31]:
autos = autos[autos["price"].between(100,350000)]
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  powerPS   model  odometer_km  registration_month fuel_type  \
0    manuell      158  andere       150000                   3       lpg   
1  automatik      286     7er       150000                   6    benzin   
2    manuell      102    golf        70000                   7    benzin   
3  automatik       71  fortwo        70000                   6    benzin   
4    manuell        0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [34]:
# Let's look at this new price data.

price = autos["price"]
price_uniq = price.unique().shape
print(price_uniq)
price_counts = price.value_counts()
price_counts.sort_index(ascending=True).head(15)

(2310,)


100    134
110      3
111      2
115      2
117      1
120     39
122      1
125      8
129      1
130     15
135      1
139      1
140      9
145      2
149      7
Name: price, dtype: int64

In [37]:
price.describe()

count     48224.000000
mean       5930.371433
std        9078.372762
min         100.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

Price offers a little more granularity. We can see that expensive cars are incredibly rare - with an average price of \\$5900 and 75% of all cars costing under \\$7500.

In [38]:
# Let's look at how common sales are.

date_freqs = autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_freqs = date_freqs * 100
    
date_freqs.describe()

count    34.000000
mean      2.941176
std       0.976582
min       0.138935
25%       2.988139
50%       3.274303
75%       3.490482
max       3.861148
Name: date_crawled, dtype: float64

Bear in mind this data comes from a 34-day period, so there are only 34 dates in question. The average day has 2.94 ads crawled, and even the largest day only has 3.86.

In [40]:
date_freqs

2016-03-05    2.536082
2016-03-06    1.403865
2016-03-07    3.606088
2016-03-08    3.317850
2016-03-09    3.301261
2016-03-10    3.228683
2016-03-11    3.259788
2016-03-12    3.691108
2016-03-13    1.567684
2016-03-14    3.666224
2016-03-15    3.431901
2016-03-16    2.946666
2016-03-17    3.149884
2016-03-18    1.289814
2016-03-19    3.473374
2016-03-20    3.780275
2016-03-21    3.720139
2016-03-22    3.288819
2016-03-23    3.228683
2016-03-24    2.944592
2016-03-25    3.149884
2016-03-26    3.230756
2016-03-27    3.112558
2016-03-28    3.496184
2016-03-29    3.411165
2016-03-30    3.373839
2016-03-31    3.185136
2016-04-01    3.369691
2016-04-02    3.560468
2016-04-03    3.861148
2016-04-04    3.653782
2016-04-05    1.306403
2016-04-06    0.317269
2016-04-07    0.138935
Name: date_crawled, dtype: float64

Our extreme outliers are the last two days of the dataset - again, this seems likely to be a function of the crawler used.

In [43]:
ad_freqs = autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_freqs = ad_freqs * 100
    
ad_freqs.head(20)

2015-06-11    0.002074
2015-08-10    0.002074
2015-09-09    0.002074
2015-11-10    0.002074
2015-12-05    0.002074
2015-12-30    0.002074
2016-01-03    0.002074
2016-01-07    0.002074
2016-01-10    0.004147
2016-01-13    0.002074
2016-01-14    0.002074
2016-01-16    0.002074
2016-01-22    0.002074
2016-01-27    0.006221
2016-01-29    0.002074
2016-02-01    0.002074
2016-02-02    0.004147
2016-02-05    0.004147
2016-02-07    0.002074
2016-02-08    0.002074
Name: ad_created, dtype: float64

In [45]:
ad_freqs.describe()

count    76.000000
mean      1.315789
std       1.595563
min       0.002074
25%       0.002074
50%       0.014516
75%       3.218315
max       3.886032
Name: ad_created, dtype: float64

We can see a greater variance here. A lot of this makes a tremendous amount of sense - less cars sold in our given time period will be from much further back. After all, ads posted months prior are much more likely to be have succeeded already.

In [46]:
last_freqs = autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_freqs = last_freqs * 100
    
last_freqs.head(150)

2016-03-05     0.107830
2016-03-06     0.431321
2016-03-07     0.543298
2016-03-08     0.732001
2016-03-09     0.958029
2016-03-10     1.063786
2016-03-11     1.240046
2016-03-12     2.378484
2016-03-13     0.887525
2016-03-14     1.262857
2016-03-15     1.586347
2016-03-16     1.644409
2016-03-17     2.809804
2016-03-18     0.732001
2016-03-19     1.575979
2016-03-20     2.065362
2016-03-21     2.054993
2016-03-22     2.135866
2016-03-23     1.857996
2016-03-24     1.976194
2016-03-25     1.909837
2016-03-26     1.667220
2016-03-27     1.555242
2016-03-28     2.084025
2016-03-29     2.229180
2016-03-30     2.469725
2016-03-31     2.382631
2016-04-01     2.285169
2016-04-02     2.488388
2016-04-03     2.513271
2016-04-04     2.453135
2016-04-05    12.506221
2016-04-06    22.196417
2016-04-07    13.215411
Name: last_seen, dtype: float64

This is quite sensible. A large number of ads will have been seen recently. After all, people are looking for cars - very large numbers of ads have been seen in the last few days.

In [47]:
autos["registration_year"].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Min is 1000, max is 9999. Clearly both erroneous values. Mean 2004 - average car is 12 years old, with 1999 and 2008 encompassing 50% of cars overall. Consequently excluding cars from pre-1920 (unlikely to be sold here) and cars after 2016 (for obvious reasons).

In [48]:
autos = autos[autos["registration_year"].between(1920,2016)]

In [49]:
rego = autos["registration_year"]

In [50]:
rego.value_counts(normalize=True)

2000    0.066969
2005    0.062805
1999    0.062114
2004    0.058231
2003    0.058101
          ...   
1927    0.000022
1929    0.000022
1950    0.000022
1948    0.000022
1952    0.000022
Name: registration_year, Length: 77, dtype: float64

In [51]:
brandlist = autos["brand"].value_counts(normalize=True)

In [52]:
brandlist[1:10]

brandy = ["bmw", "opel", "mercedes_benz", "audi", "ford", "renault", "peugeot", "fiat", 'seat']          

In [53]:
aggregate_dict = {}

mean_price = {}

for i in brandy:
    selected_rows = autos[autos["brand"] == i]
    selected_row = selected_rows.sort_values("price", ascending = False)
    aggregate_dict[i] = selected_row
    mean_price[i] = selected_row["price"].mean()


        

In [54]:
mean_price

{'bmw': 8381.677305658899,
 'opel': 3006.0002012072437,
 'mercedes_benz': 8672.654241071428,
 'audi': 9380.718547986076,
 'ford': 3779.269076305221,
 'renault': 2496.940394314535,
 'peugeot': 3113.860549132948,
 'fiat': 2836.8736310025274,
 'seat': 4433.419621749409}

We see some mostly expected results. Luxury brands like BMW and Mercedes are more expensive, Ford and Renault are less so.

In [55]:
aggregate_dict2 = {}

mean_mileage = {}

for i in brandy:
    selected_rows = autos[autos["brand"] == i]
    selected_row = selected_rows.sort_values("odometer_km", ascending = False)
    aggregate_dict2[i] = selected_row
    mean_mileage[i] = selected_row["odometer_km"].mean()


In [56]:
mean_mileage

{'bmw': 132695.32014881534,
 'opel': 129380.28169014085,
 'mercedes_benz': 131025.66964285714,
 'audi': 129245.40029835903,
 'ford': 124277.10843373495,
 'renault': 128337.91838606144,
 'peugeot': 127127.8901734104,
 'fiat': 116950.29486099411,
 'seat': 121536.64302600473}

Let's look at the mileage and compare it to the price.

In [58]:
mileage_series = pd.Series(mean_mileage)

price_series = pd.Series(mean_price)

In [59]:
mileage_and_price = pd.DataFrame(price_series, columns=['mean_price'])

mileage_and_price["mean_mileage"] = mileage_series

mileage_and_price

mean_price   mean_mileage
bmw            8381.677306  132695.320149
opel           3006.000201  129380.281690
mercedes_benz  8672.654241  131025.669643
audi           9380.718548  129245.400298
ford           3779.269076  124277.108434
renault        2496.940394  128337.918386
peugeot        3113.860549  127127.890173
fiat           2836.873631  116950.294861
seat           4433.419622  121536.643026

This is interesting. Mileage is fairly similar among all car brands, but price varies a great deal. This suggests one of the primary determinants of price is branding (or car quality, or some other quality not measured in the dataset numerically) rather than simple mileage